test/unit/org/apache/cassandra/io/sstable/format/GlobalTidyConcurrencyTest.java

```diff
@@ -0,0 +1,152 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.io.sstable.format;
+
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
+import java.util.stream.IntStream;
+
+import com.google.common.util.concurrent.Uninterruptibles;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.SequenceBasedSSTableUniqueIdentifier;
+import org.apache.cassandra.io.sstable.format.big.BigFormat;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.utils.concurrent.Ref;
+import org.jboss.byteman.contrib.bmunit.BMRule;
+import org.jboss.byteman.contrib.bmunit.BMUnitRunner;
+
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+
+@RunWith(BMUnitRunner.class)
+public class GlobalTidyConcurrencyTest
+{
+    private static final Logger logger = LoggerFactory.getLogger(GlobalTidyConcurrencyTest.class);
+
+    public static volatile boolean leakHappened = false;
+
+    @BeforeClass
+    public static void setup()
+    {
+        System.setProperty("cassandra.debugrefcount", "true");
+        System.setProperty("org.jboss.byteman.verbose", "true");
+    }
+
+    @Test
+    @BMRule(name = "Count leaks",
+    targetClass="Ref$State",
+    targetMethod="reportLeak",
+    targetLocation="AT EXIT",
+    action="org.apache.cassandra.io.sstable.format.GlobalTidyConcurrencyTest.leakHappened = true;")
+    public void tidyVsGetRaceTest() throws InterruptedException
+    {
+        int NUM_THREADS = 32;
+        Descriptor desc = createDescriptor();
+        AtomicBoolean exit = new AtomicBoolean(false);
+        CopyOnWriteArrayList<Object> errors = new CopyOnWriteArrayList<>();
+
+        class TestThread extends Thread
+        {
+            public TestThread(int idx)
+            {
+                super("test-thread-" + idx);
+            }
+
+            @Override
+            public void run()
+            {
+                try
+                {
+                    while (!exit.get())
+                    {
+                        Ref<SSTableReader.GlobalTidy> ref = SSTableReader.GlobalTidy.get(desc);
+                        LockSupport.parkNanos(ThreadLocalRandom.current().nextInt(1000000));
```

Comment:
Q: What is the difference using `parkNanos` here vs `sleep`?

Reply:
better granularity and no interrupts (at the expense of spurious wakeups which I fine with)